### ISSUE TITLE
right hand of operator incorrectly takes precedence

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#1154]
### Summary

```py
class A:
    def __or__(self, other: object) -> str:
        return ""

class B(A):
    def __ror__(self, other: object) -> int:
        return 1

class C(A):
    pass
    
def f(
    a: A = C(),
    b: B = B(),
):
    x = a | b  # ty: int, runtime: str

f()
```

this unsafety is caused by this issue:
- #630

**Comment by carljm** [comment]
Pyright gives `str` here (though it's not clear to me what logic it uses to reach that conclusion); mypy and ty both give `int`. Unlike the case in #630, mypy does not give an "unsafe overlapping dunder definitions" diagnostic here.

This seems sufficiently similar to #630 that I'm not sure it makes sense to have two separate issues open?

**Comment by KotlinIsland** [comment]
pyright gives`str` because it doesn't special case the right-hand subtype rule like mypy (and seemingly ty) does. it's invalid to take this rule into account because you don't actually know if the right is a subclass of the left, as demonstrated in the op

**Comment by KotlinIsland** [comment]
@carljm these are two separate issues, this issue can be reproduced with completely valid operator definitions, i've updated the OP to demonstrate

the issue is that the right hand subtype rule should only be applied when the **exact** types of the values are known, if it's potentially a subtype then it's unsound to do this specialisation

if we were to introduce `Exact` types, like "use-site final", or an equal lower and upper bound on a type, then we could perform this safely

**Comment by carljm** [comment]
Ok -- and the idea is that we can achieve soundness here by a) only applying the RHS subtype precedence rule if we know it applies, due to finality of the LHS, and b) implementing the warning in #630, so we would warn in any case where not applying the RHS subtype precedence rule could lead to unsound results.

Don't have time to work through all the cases in detail right now, but I think this makes sense.

**Comment by KotlinIsland** [comment]
thank you 🙂

i agree with your points. although for completeness, we should say "the lower bound of the lhs", as finality is only a single possible case

**Comment by carljm** [comment]
I think https://github.com/astral-sh/ruff/pull/26474 effectively fixed this, though in the ambiguous cases we do not emit a diagnostic, we infer the union of possible results. I think the diagnostic could make sense here, but that is tracked by #630, so I'll consider this closed.
